File: sdk/AuthSys-JS-Example/src/authsys.js
/**
 * AuthSys JavaScript SDK
 *
 * Professional authentication SDK for Node.js applications.
 */

const axios = require('axios');
const crypto = require('crypto');

class AuthSysException extends Error {
    constructor(message, statusCode = 0, errorCode = '') {
        super(message);
        this.name = 'AuthSysException';
        this.statusCode = statusCode;
        this.errorCode = errorCode;
    }
}

class AuthSys {
    constructor(options = {}) {
        this._options = {
            appSecret: options.appSecret || options.secret || '',
            appName: options.appName || options.name || '',
            version: options.version || '',
            apiUrl: (options.apiUrl || 'https://api.authsys.dpdns.org/api/v1').replace(/\/$/, ''),
            timeout: options.timeout || 30000,
            maxRetries: options.maxRetries || 3,
            skipCertificateValidation: options.skipCertificateValidation || false,
            enableLogging: options.enableLogging || false,
            hwid: options.hwid || '',
        };

        this._sessionToken = '';
        this._initialized = false;
        this._appVariables = {};
        this._username = '';
        this._heartbeatInterval = null;
    }

    _log(message) {
        if (this._options.enableLogging) {
            console.log(`[AuthSys] ${message}`);
        }
    }

    _getHwid() {
        return this._options.hwid || getHwid();
    }

    async _sendRequest(endpoint, data = null, headers = {}) {
        const url = `${this._options.apiUrl}/client/${endpoint}`;
        const reqHeaders = { 'Content-Type': 'application/json', ...headers };

        let lastError = null;
        for (let attempt = 0; attempt <= this._options.maxRetries; attempt++) {
            try {
                this._log(`POST ${url} (attempt ${attempt + 1})`);
                const config = {
                    headers: reqHeaders,
                    timeout: this._options.timeout,
                    httpsAgent: this._options.skipCertificateValidation
                        ? new (require('https').Agent)({ rejectUnauthorized: false })
                        : undefined,
                };

                const response = await axios.post(url, data, config);
                this._log(`Response: ${response.status} - ${JSON.stringify(response.data)}`);
                return response.data;
            } catch (error) {
                if (error.response) {
                    return this._handleError(error.response);
                }
                lastError = error.message;
                this._log(`Request error (attempt ${attempt + 1}): ${lastError}`);
                if (attempt < this._options.maxRetries) {
                    await new Promise(r => setTimeout(r, Math.pow(2, attempt) * 1000));
                }
            }
        }

        throw new AuthSysException(lastError || 'Request failed after all retries', 0, 'network_error');
    }

    _handleError(response) {
        let detail = response.data;
        if (typeof response.data === 'object' && response.data !== null && response.data.detail) {
            detail = response.data.detail;
        }

        const errorMap = {
            401: 'unauthorized',
            403: 'forbidden',
            404: 'not_found',
            429: 'rate_limited',
            503: 'maintenance',
        };
        const errorCode = errorMap[response.status] || 'api_error';
        throw new AuthSysException(detail, response.status, errorCode);
    }

    async init() {
        this._log('Initializing...');
        const data = {
            app_secret: this._options.appSecret,
            version: this._options.version,
            app_name: this._options.appName,
            hwid: this._getHwid(),
        };

        const result = await this._sendRequest('init', data);
        const status = result.status || '';

        if (status === 'update_required') {
            throw new AuthSysException(result.message || 'Update required', 0, 'version_mismatch');
        }

        this._initialized = status === 'success' || status === 'update_available';
        this._appVariables = result.variables || {};
        return result;
    }

    async register(username, password, licenseKey, email = '') {
        if (!this._initialized) {
            throw new AuthSysException('Not initialized. Call init() first.', 0, 'not_initialized');
        }

        const data = {
            app_secret: this._options.appSecret,
            username,
            password,
            license_key: licenseKey,
            hwid: this._getHwid(),
        };
        if (email) data.email = email;

        return await this._sendRequest('register', data);
    }

    async login(username, password, sessionLength = 86400) {
        if (!this._initialized) {
            throw new AuthSysException('Not initialized. Call init() first.', 0, 'not_initialized');
        }

        this._sessionToken = '';
        const data = {
            app_secret: this._options.appSecret,
            username,
            password,
            hwid: this._getHwid(),
            session_length: sessionLength,
        };

        const result = await this._sendRequest('login', data);
        if (result.success && result.token) {
            this._sessionToken = result.token;
            this._username = result.username || username;
        }
        return result;
    }

    async licenseLogin(licenseKey, sessionLength = 86400) {
        if (!this._initialized) {
            throw new AuthSysException('Not initialized. Call init() first.', 0, 'not_initialized');
        }

        this._sessionToken = '';
        const data = {
            app_secret: this._options.appSecret,
            license_key: licenseKey,
            hwid: this._getHwid(),
            session_length: sessionLength,
        };

        const result = await this._sendRequest('license-login', data);
        if (result.success && result.token) {
            this._sessionToken = result.token;
            this._username = result.username || '';
        }
        return result;
    }

    async licenseCheck(licenseKey) {
        const data = {
            app_secret: this._options.appSecret,
            license_key: licenseKey,
        };
        return await this._sendRequest('license/check', data);
    }

    async verify() {
        if (!this._sessionToken) {
            throw new AuthSysException('No active session. Login first.', 0, 'no_session');
        }

        const headers = {
            'Authorization': `Bearer ${this._sessionToken}`,
            'X-HWID': this._getHwid(),
        };
        return await this._sendRequest('verify', null, headers);
    }

    async sendChatMessage(roomId, message) {
        if (!this._sessionToken) {
            throw new AuthSysException('No active session. Login first.', 0, 'no_session');
        }

        const headers = { 'Authorization': `Bearer ${this._sessionToken}`, 'X-HWID': this._getHwid() };
        const endpoint = `chat/send?room_id=${roomId}&message=${encodeURIComponent(message)}`;
        return await this._sendRequest(endpoint, null, headers);
    }

    async registerDevice(hwid, deviceName = '') {
        const data = {
            app_secret: this._options.appSecret,
            hwid,
        };
        if (deviceName) data.device_name = deviceName;
        return await this._sendRequest('device/register', data);
    }

    async checkDevice(hwid) {
        const data = {
            app_secret: this._options.appSecret,
            hwid,
        };
        return await this._sendRequest('device/check', data);
    }

    getVariable(key) {
        return this._appVariables[key];
    }

    getAllVariables() {
        return this._appVariables;
    }

    logout() {
        this._sessionToken = '';
        if (this._heartbeatInterval) {
            clearInterval(this._heartbeatInterval);
            this._heartbeatInterval = null;
        }
    }

    startHeartbeat(intervalMs = 60000) {
        if (this._heartbeatInterval) clearInterval(this._heartbeatInterval);
        this._heartbeatInterval = setInterval(async () => {
            if (!this._sessionToken) {
                clearInterval(this._heartbeatInterval);
                return;
            }
            try {
                const result = await this.verify();
                if (!result.valid) {
                    this._sessionToken = '';
                    clearInterval(this._heartbeatInterval);
                }
            } catch (e) {
                this._sessionToken = '';
                clearInterval(this._heartbeatInterval);
            }
        }, intervalMs);
    }

    get isAuthenticated() {
        return !!this._sessionToken;
    }

    get isInitialized() {
        return this._initialized;
    }

    get username() {
        return this._username;
    }
}

function getHwid() {
    const os = require('os');
    const { execSync } = require('child_process');

    try {
        if (os.platform() === 'win32') {
            const output = execSync('wmic csproduct get uuid', { encoding: 'utf-8' });
            const lines = output.split('\n').filter(l => l.trim());
            if (lines.length > 1) return lines[1].trim();
        } else if (os.platform() === 'linux') {
            const fs = require('fs');
            if (fs.existsSync('/etc/machine-id')) {
                return fs.readFileSync('/etc/machine-id', 'utf-8').trim();
            }
        } else if (os.platform() === 'darwin') {
            const output = execSync('ioreg -rd1 -c IOPlatformExpertDevice', { encoding: 'utf-8' });
            const parts = output.split('"');
            if (parts.length > 3) return parts[3];
        }
    } catch (e) {}
    return os.hostname();
}

function hashString(input) {
    return crypto.createHash('sha256').update(input).digest('hex');
}

function generateGuid() {
    return crypto.randomUUID();
}

module.exports = { AuthSys, AuthSysException, getHwid, hashString, generateGuid };

File: sdk/AuthSys-Go-Example/authsys/authsys.go
package authsys

import (
	"bytes"
	"crypto/tls"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"os/exec"
	"runtime"
	"strings"
	"time"
)

type AuthSysException struct {
	Message    string
	StatusCode int
	ErrorCode  string
}

func (e *AuthSysException) Error() string {
	return fmt.Sprintf("[%s] %s", e.ErrorCode, e.Message)
}

type Options struct {
	AppSecret              string
	AppName                string
	Version                string
	ApiUrl                 string
	Timeout                int
	MaxRetries             int
	SkipCertificateValidation bool
	EnableLogging          bool
}

type AuthSys struct {
	options       *Options
	sessionToken  string
	initialized   bool
	appVariables  map[string]interface{}
	username      string
}

func NewAuthSys(appSecret string) *AuthSys {
	return &AuthSys{
		options: &Options{
			AppSecret: appSecret,
			AppName:   "",
			Version:   "",
			ApiUrl:    "https://api.authsys.dpdns.org/api/v1",
			Timeout:   30,
			MaxRetries: 3,
			SkipCertificateValidation: false,
			EnableLogging: false,
		},
		appVariables: make(map[string]interface{}),
	}
}

func NewAuthSysWithOptions(opts *Options) *AuthSys {
	if opts.ApiUrl == "" {
		opts.ApiUrl = "https://api.authsys.dpdns.org/api/v1"
	}
	if opts.Timeout == 0 {
		opts.Timeout = 30
	}
	if opts.MaxRetries == 0 {
		opts.MaxRetries = 3
	}
	return &AuthSys{
		options:      opts,
		appVariables: make(map[string]interface{}),
	}
}

func (a *AuthSys) log(message string) {
	if a.options.EnableLogging {
		fmt.Printf("[AuthSys] %s\n", message)
	}
}

func (a *AuthSys) sendRequest(endpoint string, data interface{}, headers map[string]string) (map[string]interface{}, error) {
	url := fmt.Sprintf("%s/client/%s", strings.TrimRight(a.options.ApiUrl, "/"), endpoint)
	var jsonData []byte
	if data != nil {
		jsonData, _ = json.Marshal(data)
	}

	var lastError error
	for attempt := 0; attempt <= a.options.MaxRetries; attempt++ {
		a.log(fmt.Sprintf("POST %s (attempt %d)", url, attempt+1))

		req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
		if err != nil {
			lastError = err
			continue
		}
		req.Header.Set("Content-Type", "application/json")
		for k, v := range headers {
			req.Header.Set(k, v)
		}

		client := &http.Client{Timeout: time.Duration(a.options.Timeout) * time.Second}
		if a.options.SkipCertificateValidation {
			client.Transport = &http.Transport{
				TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
			}
		}

		resp, err := client.Do(req)
		if err != nil {
			lastError = err
			a.log(fmt.Sprintf("Request error (attempt %d): %s", attempt+1, err.Error()))
			if attempt < a.options.MaxRetries {
				time.Sleep(time.Duration(1<<uint(attempt)) * time.Second)
			}
			continue
		}
		defer resp.Body.Close()

		body, _ := io.ReadAll(resp.Body)
		a.log(fmt.Sprintf("Response: %d - %s", resp.StatusCode, string(body)))

		var result map[string]interface{}
		json.Unmarshal(body, &result)

		if resp.StatusCode < 200 || resp.StatusCode >= 300 {
			detail := string(body)
			if d, ok := result["detail"]; ok {
				detail = fmt.Sprintf("%v", d)
			}
			errorCode := "api_error"
			switch resp.StatusCode {
			case 401:
				errorCode = "unauthorized"
			case 403:
				errorCode = "forbidden"
			case 404:
				errorCode = "not_found"
			case 429:
				errorCode = "rate_limited"
			case 503:
				errorCode = "maintenance"
			}
			return result, &AuthSysException{Message: detail, StatusCode: resp.StatusCode, ErrorCode: errorCode}
		}

		return result, nil
	}

	return nil, &AuthSysException{Message: fmt.Sprintf("Request failed after all retries: %v", lastError), StatusCode: 0, ErrorCode: "network_error"}
}

func (a *AuthSys) Init() (map[string]interface{}, error) {
	a.log("Initializing...")
	data := map[string]interface{}{
		"app_secret": a.options.AppSecret,
		"version":    a.options.Version,
		"app_name":   a.options.AppName,
		"hwid":       GetHwid(),
	}

	result, err := a.sendRequest("init", data, nil)
	if err != nil {
		return result, err
	}

	status := ""
	if s, ok := result["status"].(string); ok {
		status = s
	}

	if status == "update_required" {
		msg := ""
		if m, ok := result["message"].(string); ok {
			msg = m
		}
		return result, &AuthSysException{Message: msg, StatusCode: 0, ErrorCode: "version_mismatch"}
	}

	a.initialized = status == "success" || status == "update_available"
	if vars, ok := result["variables"].(map[string]interface{}); ok {
		a.appVariables = vars
	}
	return result, nil
}

func (a *AuthSys) Register(username, password, licenseKey, email string) (map[string]interface{}, error) {
	if !a.initialized {
		return nil, &AuthSysException{Message: "Not initialized. Call Init() first.", StatusCode: 0, ErrorCode: "not_initialized"}
	}

	data := map[string]interface{}{
		"app_secret":   a.options.AppSecret,
		"username":     username,
		"password":     password,
		"license_key":  licenseKey,
		"hwid":         GetHwid(),
	}
	if email != "" {
		data["email"] = email
	}

	return a.sendRequest("register", data, nil)
}

func (a *AuthSys) Login(username, password string, sessionLength int) (map[string]interface{}, error) {
	if !a.initialized {
		return nil, &AuthSysException{Message: "Not initialized. Call Init() first.", StatusCode: 0, ErrorCode: "not_initialized"}
	}

	a.sessionToken = ""
	data := map[string]interface{}{
		"app_secret":     a.options.AppSecret,
		"username":       username,
		"password":       password,
		"hwid":           GetHwid(),
		"session_length": sessionLength,
	}

	result, err := a.sendRequest("login", data, nil)
	if err != nil {
		return result, err
	}

	if token, ok := result["token"].(string); ok && token != "" {
		a.sessionToken = token
	}
	if u, ok := result["username"].(string); ok {
		a.username = u
	}
	return result, nil
}

func (a *AuthSys) LicenseLogin(licenseKey string, sessionLength int) (map[string]interface{}, error) {
	if !a.initialized {
		return nil, &AuthSysException{Message: "Not initialized. Call Init() first.", StatusCode: 0, ErrorCode: "not_initialized"}
	}

	a.sessionToken = ""
	data := map[string]interface{}{
		"app_secret":     a.options.AppSecret,
		"license_key":    licenseKey,
		"hwid":           GetHwid(),
		"session_length": sessionLength,
	}

	result, err := a.sendRequest("license-login", data, nil)
	if err != nil {
		return result, err
	}

	if token, ok := result["token"].(string); ok && token != "" {
		a.sessionToken = token
	}
	if u, ok := result["username"].(string); ok {
		a.username = u
	}
	return result, nil
}

func (a *AuthSys) LicenseCheck(licenseKey string) (map[string]interface{}, error) {
	data := map[string]interface{}{
		"app_secret":  a.options.AppSecret,
		"license_key": licenseKey,
	}
	return a.sendRequest("license/check", data, nil)
}

func (a *AuthSys) Verify() (map[string]interface{}, error) {
	if a.sessionToken == "" {
		return nil, &AuthSysException{Message: "No active session. Login first.", StatusCode: 0, ErrorCode: "no_session"}
	}

	headers := map[string]string{
		"Authorization": fmt.Sprintf("Bearer %s", a.sessionToken),
		"X-HWID":        GetHwid(),
	}
	return a.sendRequest("verify", nil, headers)
}

func (a *AuthSys) SendChatMessage(roomId int, message string) (map[string]interface{}, error) {
	if a.sessionToken == "" {
		return nil, &AuthSysException{Message: "No active session. Login first.", StatusCode: 0, ErrorCode: "no_session"}
	}

	headers := map[string]string{
		"Authorization": fmt.Sprintf("Bearer %s", a.sessionToken),
		"X-HWID":        GetHwid(),
	}
	endpoint := fmt.Sprintf("chat/send?room_id=%d&message=%s", roomId, url.QueryEscape(message))
	return a.sendRequest(endpoint, nil, headers)
}

func (a *AuthSys) RegisterDevice(hwid, deviceName string) (map[string]interface{}, error) {
	data := map[string]interface{}{
		"app_secret": a.options.AppSecret,
		"hwid":       hwid,
	}
	if deviceName != "" {
		data["device_name"] = deviceName
	}
	return a.sendRequest("device/register", data, nil)
}

func (a *AuthSys) CheckDevice(hwid string) (map[string]interface{}, error) {
	data := map[string]interface{}{
		"app_secret": a.options.AppSecret,
		"hwid":       hwid,
	}
	return a.sendRequest("device/check", data, nil)
}

func (a *AuthSys) GetVariable(key string) interface{} {
	return a.appVariables[key]
}

func (a *AuthSys) GetAllVariables() map[string]interface{} {
	return a.appVariables
}

func (a *AuthSys) Logout() {
	a.sessionToken = ""
}

func (a *AuthSys) IsAuthenticated() bool {
	return a.sessionToken != ""
}

func (a *AuthSys) IsInitialized() bool {
	return a.initialized
}

func (a *AuthSys) GetUsername() string {
	return a.username
}
